### PR TITLE
Adds new list and loop. [Fixes #2822]

### LINKED ISSUE #2822: Update "Ask users for... Multiple Responses pattern"
### Duplicate check

- [X] I've searched for any related issues and avoided creating a duplicate issue.

### This update is for:

Pattern

### What is the name?

Ask users for... Multiple responses

### What is the nature of this update?

- [ ] How to build this component/pattern
- [X] When to use this component/pattern
- [X] When to use something else
- [X] Usability guidance
- [X] Accessibility considerations
- [X] Content considerations
- [X] Implementation
- [ ] Package information
- [ ] Addition to Forms Library documentation
- [ ] Update to existing Forms Library documentation

### What problem does this solve?

The forms team has added a reusable pattern for Multiple responses List and Loop. This allows team to consistently and repeatably build a list and loop pattern that has accessibility baked in and all of the necessary components/modals/etc.

This particular variation builds a multiple response pattern utilizing "one thing per page".

I did find a number of tickets related to the Multiple responses pattern. I don't think this particular work overlaps with any of them, but I'm listing them here for reference:

https://github.com/department-of-veterans-affairs/vets-design-system-documentation/issues/2562
https://github.com/department-of-veterans-affairs/vets-design-system-documentation/issues/1697
https://github.com/department-of-veterans-affairs/vets-design-system-documentation/issues/1187
https://github.com/department-of-veterans-affairs/vets-design-system-documentation/issues/1565

### Here are some assets that would be useful:
1. [Figma file](https://www.figma.com/design/OxzngjwHg4n0oNSGHkwJ5Z/DSC---WIP---Pattern---Multi-page-List-and-Loop?node-id=1-128050&t=dUIefNfc9EA0q6ag-11) (Working with Dan and Lucas to get this moved over to Design System space) 
2. [ReadMe file](https://github.com/department-of-veterans-affairs/vets-website/blob/main/src/platform/forms-system/src/js/patterns/array-builder/README.md)
3. [Mock Form Patterns](https://staging.va.gov/mock-form-patterns/introduction) contains two examples of the pattern
4. [Mural lo-fi user flow](https://app.mural.co/t/departmentofveteransaffairs9999/m/departmentofveteransaffairs9999/1715783901077/5f9acc74ad8555d4219a14eaf9836c46dbb8dac4?sender=ud94436ff2bf5d5ce166a5253)

### Content that might be useful to use/incorporate

# How to design and build - Multi-page (Multiple responses)

Use this pattern when users need to add similar information multiple times, such as information about dependents.  This method allows more table-like data to be collected following the [One thing per page](https://design.va.gov/patterns/ask-users-for/a-single-response#what-is-one-thing-per-page) pattern. 

**Resources** 
- [Mock Form Patterns](https://staging.va.gov/mock-form-patterns/introduction) contains two examples of the pattern, and you can test them out now!
- The [ReadMe file](https://github.com/department-of-veterans-affairs/vets-website/blob/main/src/platform/forms-system/src/js/patterns/array-builder/README.md) contains instructions for setting up the scaffolding for this pattern in just a few minutes.
- A [Figma template](https://www.figma.com/file/4A3O3mVx4xDAKfHE7fPF1U/VADS-Templates%2C-Patterns%2C-and-Forms?type=design&node-id=5751%3A10777&mode=design&t=FYk7L3PJ9a16WGP9-1) with all of the necessary components is available. 

## How it works

We're following the [One thing per page](#) pattern. By chunking information in smaller sets, users are able to get through the form with greater ease.

image

1. A user starts enters either a required or optional list and loop flow.
    **- A required item for the list and loop:**
    The required version has an intro page explaining that we're about to ask a series of questions about a thing, and if there's a limit, they can enter up to that limit. Teams can customize how they want this page to work.

    **- Completely optional list and loop:**
    For an optional list and loop, the question is asked "Do you have [a thing] to add? yes/no" Yes enters the loop.

2. The first page of the list and loop asks them to identify the thing they are adding. In this case, treatment records from a facility. Designers can choose how they want to construct the pages of the list and loop. We recommend following the one thing per page pattern as we do throughout our forms.
3. At any time in the multiple responses list and loop a user can exit the loop by clicking the "cancel adding this [thing]" button.  A modal appears asking them to confirm their choice. If they answer that they want to stop adding this thing, any data they have entered is removed, and the user is returned to the first page of the list and loop pattern.
4. Summary list page allows uers to review what they've entered, edit, delete items, or add another. 
5. If the user chooses to add another item they return to the first question page to add the new item.

The loop can be repeated as many times as your form allows.
image

## Summary list page

image

After adding each item, the user is shown a summary of what they have added so far and they can

- **Edit items:** Clicking the edit link returns the user to the first question of the loop and the user has now entered an "edit flow". After editing items, the user is returned to the summary page and alert is shown confirming their item has been updated.  When entering an edit flow, the H3s of the pages are updated to include "Edit [previous h3 title]"
- **Delete items:** Clicking the delete button, allows a user to confirm that they want to delete the entire loop and all of the answers they had given through the entire loop. A modal pops up asking the user to confirm their choice. After they remove a card, they are returned to the summary page where they can choose to add another. IF the user has removed all items, and yet one item is required, the user is returned to the first page of the loop where a warning alert is displayed reminding the user at least one item is required.
- **Add another item:** If the user chooses to add another item they return to the first question page to add the new item.
- Continue to the next question in the form

These summary cards work the same on the review page of the form.

## Content

The pattern has baked in error and validation messages for:
- Successful editing of an item
- Successful removal of an item
- When the maximum number of items have been added
- When all items in a required loop have been removed
- When a user wants to cancel adding an item mid-flow

The pattern has also baked patterns for using the same word for adding an item on question pages, summary cards, and edit pages. For example:
- Do you want to add another [dependent]?
- Review your [dependents]
- Remove a [dependent]
- You have added the maximum number of [dependents] 

## Other things to consider

If at least one item is required, use hint text to let users know at least one item is required. If all items are removed, return users to the first page of the loop to gather information.

If there is a maximum number of items, make this clear to the user. You can use hint text to do this. Also, after the user has entered the maximum number allowed the pattern removes the "add another" question, and displays a warning instructing the user that they have entered the maximum allowed and they can either edit, or remove a card if they need to add more information.

## Accessibility
On the summary page, ensure that the edit links and delete buttons have accessible text so that screen reader users understand what is being edited or deleted.

### PR DESCRIPTION
**Pull request by humancompanion-usds**

### CODE REVIEW

**Review by Andrew565 — APPROVED** [review]

File: src/_patterns/ask-users-for/multiple-responses.md [hunk 1]
```diff
@@ -33,59 +36,102 @@ anchors:
...
+3. At any time in the multiple responses list and loop a user can exit the loop by clicking the "cancel adding this [thing]" button. A modal appears asking them to confirm their choice. If they answer that they want to stop adding this thing, any data they have entered is removed, and the user is returned to the first page of the list and loop pattern.
+4. Summary list page allows users to review what they've entered, edit, delete items, or add another.
+5. If the user chooses to add another item they return to the first question page to add the new item.
 
-### Review for edit and remove
+The loop can be repeated as many times as your form allows.
 
-{% include component-example.html alt="Form showing service history collected from a Veteran." file="/images/patterns/ask-users-for/multiple-responses/adding-service-period.png" caption="Form displaying service history collected from a Veteran and allowing the user to edit or remove the information collected." class="x2" %}
+<a class="vads-c-action-link--blue" href="{{ page.example-link-multi-page-requried }}">
```
Comment (danbrady) [inline src/_patterns/ask-users-for/multiple-responses.md:57]:
```suggestion
<a class="vads-c-action-link--blue" href="{{ page.example-link-multi-page-required }}">
```

Comment (danbrady) [reply src/_patterns/ask-users-for/multiple-responses.md:57]: Link has typo.

File: src/_patterns/ask-users-for/multiple-responses.md [hunk 2]
```diff
@@ -33,59 +36,102 @@ anchors:
...
 
-### When to use this variation
+After adding each item, the user is shown a summary of what they have added so far and they can:
 
-This variation was the first to be created and thus is considered the default and is the most widely in use.
+* **Edit items.** Clicking the edit link returns the user to the first question of the loop and the user has now entered an "edit flow". After editing items, the user is returned to the summary page and alert is shown confirming their item has been updated. When entering an edit flow, the H3s of the pages are updated to include "Edit [previous h3 title]"
+* **Delete items.** Clicking the delete button allows a user to confirm that they want to delete the entire loop and all of the answers they have given through the entire loop. A modal pops up asking the user to confirm their choice. After they remove a card, they are returned to the summary page where they can choose to add another. IF the user has removed all items, and yet one item is required, the user is returned to the first page of the loop where a warning alert is displayed reminding the user at least one item is required.
```
Comment (danbrady) [inline src/_patterns/ask-users-for/multiple-responses.md:69]:
```suggestion
* **Delete items.** Clicking the delete button allows a user to confirm that they want to delete the entire loop and all of the answers they have given through the entire loop. A modal pops up asking the user to confirm their choice. After they remove a card, they are returned to the summary page where they can choose to add another. If the user has removed all items, and yet one item is required, the user is returned to the first page of the loop where a warning alert is displayed reminding the user at least one item is required.
```

Is this purposely capitalized for emphasis?

File: src/_patterns/ask-users-for/multiple-responses.md [hunk 3]
```diff
@@ -33,59 +36,102 @@ anchors:
...
-* The "Add another" button stays visible so that additional entries can be made.
+{% include component-example.html alt="A summary page for the multiple response multi-page pattern variation." file="/images/patterns/ask-users-for/multiple-responses/multiple-response-summary.png" caption="An example of the summary page for a multi-page multiple response pattern. This summary page reflects the data collected thus far and allows the user to act on that data or add more." class="x2" %}
 
-### When to use this variation
+After adding each item, the user is shown a summary of what they have added so far and they can:
 
-This variation was the first to be created and thus is considered the default and is the most widely in use.
+* **Edit items.** Clicking the edit link returns the user to the first question of the loop and the user has now entered an "edit flow". After editing items, the user is returned to the summary page and alert is shown confirming their item has been updated. When entering an edit flow, the H3s of the pages are updated to include "Edit [previous h3 title]"
```
Comment (danbrady) [inline src/_patterns/ask-users-for/multiple-responses.md:68]:
```suggestion
* **Edit items.** Clicking the edit link returns the user to the first question of the loop and the user has now entered an "edit flow". After editing items, the user is returned to the summary page and alert is shown confirming their item has been updated. When entering an edit flow, the H3s of the pages are updated to include "Edit [previous h3 title]".
```

File: src/_patterns/ask-users-for/multiple-responses.md [hunk 4]
```diff
@@ -33,59 +36,102 @@ anchors:
...
+* **Delete items.** Clicking the delete button allows a user to confirm that they want to delete the entire loop and all of the answers they have given through the entire loop. A modal pops up asking the user to confirm their choice. After they remove a card, they are returned to the summary page where they can choose to add another. IF the user has removed all items, and yet one item is required, the user is returned to the first page of the loop where a warning alert is displayed reminding the user at least one item is required.
+* **Add another item.** If the user chooses to add another item they return to the first question page to add the new item.
+Continue to the next question in the form
```
Comment (danbrady) [inline src/_patterns/ask-users-for/multiple-responses.md:71]:
```suggestion
* **Continue to the next question in the form**
```

I think this is intended to be separate bulleted item?

File: src/_patterns/ask-users-for/multiple-responses.md [hunk 5]
```diff
@@ -33,59 +36,102 @@ anchors:
...
+* When a user wants to cancel adding an item mid-flow
 
-{% include component-example.html alt="Form example allowing user to build a list of items and then adding details to each item in the list." file="/images/patterns/ask-users-for/multiple-responses/array-data.png" caption="Form sub-steps that allow a Veteran to build a list of items and then add details to each item in the list." class="x2" %}
+The pattern also has baked-in patterns for using the same word for adding an item on question pages, summary cards, and edit pages. For example:
 
-This pattern is carried out in several steps within a single form step:
+* Do you want to add another [dependent]?
+* Review your [dependents]
+* Remove a [dependent]
+* You have added the maximum number of [dependents]
```
Comment (danbrady) [inline src/_patterns/ask-users-for/multiple-responses.md:98]: I feel we need a sentence after the list of examples above and before the two following bullets regarding suggestions. Currently it reads like the following two bullets are part of the examples list.

File: src/assets/sprite.svg
Comment (danbrady) [inline src/assets/sprite.svg:1]: Not sure why the sprite file is included in this PR?

Comment (humancompanion-usds) [reply src/assets/sprite.svg:1]: Yeah, I'd like to know why this sprite file keeps getting regenerated in the background as well!

File: src/_patterns/ask-users-for/multiple-responses.md (hunk 5 above)
Comment (humancompanion-usds) [reply src/_patterns/ask-users-for/multiple-responses.md]: Fixed.

**Pull request merged**